exporter/postgres/setup_pg12.sql

```diff
@@ -133,5 +133,141 @@ WHERE (ROUND(used/slots*100)::int) > p_percent;
 $function$;
 
 
+/*
+ * Table and functions for monitoring changes to pg_settings and pg_hba_file_rules system catalogs.
+ * Can't just do a raw check for the hash value since Prometheus only records numeric values for alerts
+ * Table allows recording of existing settings so it can be referred back to to see what changed
+ * If function returns 0, then NO settings have changed for either pg_settings or hba since last known valid state 
+ * If function returns 1, then just pg_settings have changed since last known valid state
+ * If function returns 2, then just hba has changed since last known valid state
+ * If function returns 3, then both pg_settings and hba have changed since last known valid state
+ */
+
+DROP TABLE IF EXISTS monitor.settings_checksum;
+DROP FUNCTION IF EXISTS monitor.settings_checksum();
+
+CREATE TABLE monitor.settings_checksum (
+    settings_hash text NOT NULL
+    , settings_string text NOT NULL
+    , hba_hash text NOT NULL
+    , hba_string text NOT NULL
+    , created_at timestamptz DEFAULT now() NOT NULL
+    , valid smallint NOT NULL );
+
+COMMENT ON COLUMN monitor.settings_checksum.valid IS 'Set this column to zero if this group of settings is a valid change';
+CREATE INDEX ON monitor.settings_checksum (created_at);
+
+CREATE FUNCTION monitor.settings_checksum () RETURNS smallint
+    LANGUAGE plpgsql SECURITY DEFINER 
+AS $function$
+DECLARE
+
+v_hba_hash              text;
+v_hba_hash_old          text;
+v_hba_match             smallint := 0;
+v_hba_string            text;
+v_hba_string_old        text;
+v_match_total           smallint := 0;
+v_settings_hash         text;
+v_settings_hash_old     text;
+v_settings_match        smallint := 0;
+v_settings_string       text;
+v_settings_string_old   text;
+v_valid                 smallint;
+
+BEGIN
+
+WITH settings_ordered_list AS (
+    SELECT name
```

Comment:
Does it makes sense to also consider pending_restart ?
Since value has not yet changed

Reply:
I'm not sure how that would fit in here. The setting itself has not changed yet, so causing the checksum monitor to alert wouldn't make sense since nothing actually changed yet and there would be nothing in the log to indicate what "would" change.

I would consider a separate metric to watch for if any settings change into a `pending_restart` state.

Reply:
Just as a note, when a setting goes into `pending_restart = true`, the setting value itself does not show as changed in pg_settings.

Reply:
true but this flag be used as warning level alert. Next restart will cause new value for this setting.

Reply:
I thought about creating another number value for pending_restart state in `ccp_settings_checksum` , but that's still going to cause the metric to alert yet again on the next restart with a different state value. So I've created a new metrics just for monitoring pending_restart state so it's not causing confusion with the checksum monitor: `ccp_settings_pending_restart`

Issue with this is that that column was only added in 9.5. On current release schedule, PG9.4 will be going out of support before the next version of pgmonitor is released. Can we drop support for that with the next version of pgmonitor?

Reply:
> Issue with this is that that column was only added in 9.5. 

pg_hba_file_rules was added in PG10 ¯\_(ツ)_/¯

Reply:
Forgot that was a 10 only feature. I also realized that this full method of change tracking only works on a primary since that's the only place a log can be written. If someone changes a replica, we can still alert, but no change log is available.

Reply:
I fixed the function so it will only report hba changes for PG10+. Allows the same function to be used on all versions. Only settings monitoring is available for 9.6 and earlier.

Also added in the ability for this to work on replicas. If the replica's config will be different than the primary, a known good hash can be passed to the function (since logs cannot be written). I provided views for users to easily be able to generate the same hashes we use.

The question now is still whether to support 9.4. If we can drop that, I can just add the `pending_restart` check to the `queries_common.yml` file. Otherwise, will have to add it individually to all files for 9.5+ then move it to common later once we do drop 9.4.

Reply:
for pre pg 10 can we create file fdw and acess this info ?

Reply:
> I thought about creating another number value for pending_restart state in `ccp_settings_checksum` , but that's still going to cause the metric to alert yet again on the next restart with a different state value. So I've created a new metrics just for monitoring pending_restart state so it's not causing confusion with the checksum monitor: `ccp_settings_pending_restart`
> 

But two alerts are different
1. Value changed
2. It might change on next restart

> Issue with this is that that column was only added in 9.5. On current release schedule, PG9.4 will be going out of support before the next version of pgmonitor is released. Can we drop support for that with the next version of pgmonitor?

So pre 9.5 we will not utilize this feature.

Reply:
The file_fdw would make this inordinately complex and may involve manual setup (not sure I can do all the server/user setup via the script). Also, this doesn't account for the use of @ options to include content from other files. I'm leaning towards this just being a 10+ feature for now and unless a client specifically requests this and is unable to upgrade.

As for it being two separate alerts, yes, that's because it's two separate things. One is a pending change to the config and the other is an actual change to the active postgresql.conf configuration. They should be two distinct alerts, IMO, to avoid the confusion. Honestly, could even make the postgresql.conf and pg_hba.conf monitors two different metrics as well. May even make it easier to manage and clearer for the users.

My request is to drop 9.4 support all together at this point for the simplification of the configuration files given above. We've already stated we were going to keep on track with the supported versions of PG.

Reply:
`ccp_pg_settings_checksum` and `ccp_pg_hba_checksum` are now two separate metrics. Wish I would've thought of that earlier because it makes the code vastly more simple as well as being clearer for the user as well I think.